book-content/chapters/11-annotations-and-assertions.md

```diff
@@ -0,0 +1,1196 @@
+
+### The Non-null Assertion
+
+Another assertion we can use is the non-null assertion, which is specified by using the `!` operator. This provides a quick way to tell TypeScript that a value is not `null` or `undefined`.
+
+Heading back to our `searchParams` example from earlier, we can use the non-null assertion to tell TypeScript that `id` will never be `null`:
+
+```typescript
+const searchParams = new URLSearchParams(window.location.search);
+
+const id = searchParams.get("id")!;
+```
+
+This forces TypeScript to treat `id` as a string, even though it could be `null` at runtime. It's the equivalent of using `as string`, but is a little more convenient.
+
+You can also use it when accessing a property which may or may not be defined:
+
+```typescript
+type User = {
+  name: string;
+  profile?: {
+    bio: string;
+  };
+};
+
+const logUserBio = (user: User) => {
+  console.log(user.profile!.bio);
+};
+```
+
+Or, when calling a function that might not be defined:
+
+```typescript
+type Logger = {
+  log?: (message: string) => void;
+};
+
+const main = (logger: Logger) => {
+  logger.log!("Hello, world!");
+};
+```
+
+Each of these fails at runtime if the value is not defined. But it's a convenient lie to TypeScript that we're sure it will be.
+
+The non-null assertion, like other assertions, is a dangerous tool. It's particularly nasty because it's one character long, so easier to miss than `as`.
+
+For fun, I like to use at least three or four in a row to make sure developers know that what they're doing is dangerous:
+
+```typescript
+// Yes, this syntax is legal
+const id = searchParams.get("id")!!!!;
+```
+
+## Error Suppression Directives
+
+Assertions are not the only ways we can lie to TypeScript. There are several comment directives that can be used to suppress errors.
+
+### `@ts-expect-error`
+
+Throughout the book's exercises we've seen several examples of `@ts-expect-error`. This directive gives us a way to tell TypeScript that we expect an error to occur on the next line of code.
+
+In this example, we're creating an error by passing a string into a function that expects a number.
+
+```typescript
+function addOne(num: number) {
+  return num + 1;
+}
+
+// @ts-expect-error
+const result = addOne("one");
+```
+
+But the error doesn't show up in the editor, because we told TypeScript to expect it.
+
+However, if we pass a number into the function, the error will show up:
+
+```ts twoslash
+// @errors: 2578
+function addOne(num: number) {
+  return num + 1;
+}
+
+// ---cut---
+// @ts-expect-error
+const result = addOne(1);
+```
+
+So, TypeScript expects every `@ts-expect-error` directive to be _used_ - to be followed by an error.
+
+Frustratingly, `@ts-expect-error` doesn't let you expect a specific error, but only that an error will occur.
```

Comment:
and it suppresses everything on the next line, so it's fairly easy to suppress more than one error this way